include/otf2xx/event/metric.hpp

```diff
@@ -267,12 +576,33 @@ namespace event
             metric_instance_ = mi;
         }
 
+        otf2::definition::metric_class resolve_metric_class() const
+        {
+            return otf2::definition::metric_class{ resolve_metric_class_ref() };
+        }
+
         friend class otf2::writer::local;
 
     private:
-        otf2::definition::detail::weak_ref<otf2::definition::metric_class> metric_class_;
-        otf2::definition::detail::weak_ref<otf2::definition::metric_instance> metric_instance_;
-        std::vector<value_container> values_;
+        template <typename Definition>
+        using weak_ref = otf2::definition::detail::weak_ref<Definition>;
+
+        weak_ref<otf2::definition::metric_class> resolve_metric_class_ref() const
+        {
+            if (has_metric_instance())
+            {
+                return otf2::definition::make_weak_ref(metric_instance_->metric_class());
+            }
+            else
+            {
+                return metric_class_;
+            }
+        }
+
+    private:
+        weak_ref<otf2::definition::metric_class> metric_class_;
+        weak_ref<otf2::definition::metric_instance> metric_instance_;
+        metric_values values_;
     };
 } // namespace event
 } // namespace otf2
```

Comment:
Please split the file metric.hpp in 3 files:

- event::metric
- event::detail::type_value_proxy
- event::detail::metric_values